refactor: mempool: use CTxMemPool::Limits

stickies-v opened the pull request:
Mempool currently considers 4 limits regarding ancestor and descendant count and size, which get passed around between functions quite a bit. This PR uses `CTxMemPool::Limits` introduced in https://github.com/bitcoin/bitcoin/pull/25290 to simplify those signatures and callsites.

The purpose of this PR is to improve readability and maintenance, without behaviour change.

As noted in the first commit "refactor: mempool: change MemPoolLimits members to uint", we currently have an underflow issue where a user could pass a negative `-limitancestorsize`, which is eventually cast to an unsigned integer. This behaviour already exists. Because it's orthogonal and to minimize scope, I think this should be fixed in a separate PR.

stickies-v committed:
refactor: mempool: use CTxMempool::Limits

Simplifies function signatures by removing repetition of all the
ancestor/descendant limits,  and increases readability by being
more verbose by naming the limits, while still reducing the LoC.

stickies-v committed:
refactor: mempool: add MemPoolLimits::NoLimits()

There are quite a few places in the codebase that require us to
construct a CTxMemPool without limits on ancestors and descendants.
This helper function allows us to get rid of all that duplication.

glozow commented on src/test/mempool_tests.cpp:
```diff
@@ -203,7 +203,13 @@ BOOST_AUTO_TEST_CASE(MempoolIndexingTest)
     CTxMemPool::setEntries setAncestorsCalculated;
     std::string dummy;
-    BOOST_CHECK_EQUAL(pool.CalculateMemPoolAncestors(entry.Fee(2000000LL).FromTx(tx7), setAncestorsCalculated, 100, 1000000, 1000, 1000000, dummy), true);
+    CTxMemPool::Limits mempool_limits{
+        .ancestor_count = 100,
+        .ancestor_size_vbytes = 1'000'000,
+        .descendant_count = 1'000,
+        .descendant_size_vbytes = 1'000'000,
+    };
+    BOOST_CHECK_EQUAL(pool.CalculateMemPoolAncestors(entry.Fee(2'000'000LL).FromTx(tx7), setAncestorsCalculated, mempool_limits, dummy), true);
```
I think this can be replaced with `CTxMemPool::Limits::NoLimits()`, actually. These numbers seem arbitrarily high and don't restrict anything, suggesting the intent of this test case is just to calculate the ancestors properly.

glozow commented on src/kernel/mempool_limits.h:
```diff
@@ -17,13 +17,21 @@ namespace kernel {
+    static MemPoolLimits NoLimits() {
+        uint64_t no_limit{std::numeric_limits<uint64_t>::max()};
+        return {no_limit, no_limit, no_limit, no_limit};
+    }
```
I think it would make sense to delete the default ctor and require callers to set all 4 members when constructing.

stickies-v replied:
What would be the benefit of that approach? Asking because it's quite a deviation from the current process, where we construct an empty MemPoolLimits (with default values as defined here) in init.cpp (as a member of CTxMemPool::Options), and then pass it to multiple AppluArgsManOptions() functions to update it with startup parameters if provided. I think deleting the default ctor of MemPoolLimits would require changing that whole flow, and I'm not sure it's worth it. I think having the default values here and not requiring every value to be explicitly defined makes sense?

glozow commented on src/kernel/mempool_limits.h:
```diff
@@ -17,13 +17,21 @@ namespace kernel {
 struct MemPoolLimits {
     //! The maximum allowed number of transactions in a package including the entry and its ancestors.
-    int64_t ancestor_count{DEFAULT_ANCESTOR_LIMIT};
+    uint64_t ancestor_count{DEFAULT_ANCESTOR_LIMIT};
```
Can you explain why the change from signed to unsigned?

Riahiamirreza replied:
I'm new to the code base so my question may look silly. Why the type of ancestor and descendant count are 64 bit? Isn't  32 bit more than sufficient? Although, I know that's not in the scope of this PR.

glozow replied:
> Currently, the int64_t members later on in the call stack
> get cast into a size_t

Yeah so we have virtual size implemented as size_t, unsigned ints, and signed ints in various places in the codebase, and it would be good to switch to 1 type consistently. I don't think there's a PR to do all of them, but #23962 is one.

My current thinking is to use signed ints instead of size_t because we're doing arithmetic with these limits (see `CalculateAncestorsAndCheckLimits`, https://github.com/bitcoin/bitcoin/pull/23962#issuecomment-1004994681). So I always use `int64_t` to hold virtual sizes. Not sure where we're going long term :shrug:, but that's why I don't really agree with switching to unsigned.

> Why the type of ancestor and descendant count are 64 bit? Isn't 32 bit more than sufficient?

There's never a need for more than 32 bits, but switching partially would cause integer truncation somewhere.

stickies-v replied:
I understand your reasoning and the one from the linked comment, but I'm not sure it makes sense here. A lot of the comparisons we do are with other `size_t`'s, e.g. from `std::set::size()`. So, we'd either have to unsafely cast them into an int64_t (_probably_ fine, but... not ideal?) or implement a function that does it with bounds checking etc.

I'm not sure that's preferable? See e.g. the below diff, with some unsafe static_cast as well as implicit conversion (e.g. `CheckPackageLimits()`  passing `package.size()` to `CalculateAncestorsAndCheckLimits()`)

<details>
<summary>git diff</summary>

```diff
diff --git a/src/kernel/mempool_limits.h b/src/kernel/mempool_limits.h
index fd0979c6c..584af5679 100644
--- a/src/kernel/mempool_limits.h
+++ b/src/kernel/mempool_limits.h
@@ -17,20 +17,20 @@ namespace kernel {
  */
 struct MemPoolLimits {
     //! The maximum allowed number of transactions in a package including the entry and its ancestors.
-    uint64_t ancestor_count{DEFAULT_ANCESTOR_LIMIT};
+    int64_t ancestor_count{DEFAULT_ANCESTOR_LIMIT};
     //! The maximum allowed size in virtual bytes of an entry and its ancestors within a package.
-    uint64_t ancestor_size_vbytes{DEFAULT_ANCESTOR_SIZE_LIMIT_KVB * 1'000};
+    int64_t ancestor_size_vbytes{DEFAULT_ANCESTOR_SIZE_LIMIT_KVB * 1'000};
     //! The maximum allowed number of transactions in a package including the entry and its descendants.
-    uint64_t descendant_count{DEFAULT_DESCENDANT_LIMIT};
+    int64_t descendant_count{DEFAULT_DESCENDANT_LIMIT};
     //! The maximum allowed size in virtual bytes of an entry and its descendants within a package.
-    uint64_t descendant_size_vbytes{DEFAULT_DESCENDANT_SIZE_LIMIT_KVB * 1'000};
+    int64_t descendant_size_vbytes{DEFAULT_DESCENDANT_SIZE_LIMIT_KVB * 1'000};
 
     /**
      * @return MemPoolLimits with all the limits set to the maximum
      */
     static MemPoolLimits NoLimits()
     {
-        uint64_t no_limit{std::numeric_limits<uint64_t>::max()};
+        int64_t no_limit{std::numeric_limits<int64_t>::max()};
         return {no_limit, no_limit, no_limit, no_limit};
     }
 
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index 3097473ac..ee5ffff7b 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -183,14 +183,14 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256>& vHashes
     }
 }
 
-bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
-                                                  size_t entry_count,
+bool CTxMemPool::CalculateAncestorsAndCheckLimits(int64_t entry_size,
+                                                  int64_t entry_count,
                                                   setEntries& setAncestors,
                                                   CTxMemPoolEntry::Parents& staged_ancestors,
                                                   const Limits& limits,
                                                   std::string &errString) const
 {
-    size_t totalSizeWithAncestors = entry_size;
+    int64_t totalSizeWithAncestors = entry_size;
 
     while (!staged_ancestors.empty()) {
         const CTxMemPoolEntry& stage = staged_ancestors.begin()->get();
@@ -241,7 +241,7 @@ bool CTxMemPool::CheckPackageLimits(const Package& package,
             std::optional<txiter> piter = GetIter(input.prevout.hash);
             if (piter) {
                 staged_ancestors.insert(**piter);
-                if (staged_ancestors.size() + package.size() > limits.ancestor_count) {
+                if (static_cast<int64_t>(staged_ancestors.size() + package.size()) > limits.ancestor_count) {
                     errString = strprintf("too many unconfirmed parents [limit: %u]", limits.ancestor_count);
                     return false;
                 }
@@ -277,7 +277,7 @@ bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry,
             std::optional<txiter> piter = GetIter(tx.vin[i].prevout.hash);
             if (piter) {
                 staged_ancestors.insert(**piter);
-                if (staged_ancestors.size() + 1 > limits.ancestor_count) {
+                if (static_cast<int64_t>(staged_ancestors.size() + 1) > limits.ancestor_count) {
                     errString = strprintf("too many unconfirmed parents [limit: %u]", limits.ancestor_count);
                     return false;
                 }

```

</details>

stickies-v replied:
@glozow:

> So I always use int64_t to hold virtual sizes. Not sure where we're going long term 🤷, but that's why I don't really agree with switching to unsigned.

I went through all the places where any of the 4 members of `MemPoolLimits` are accessed. In all of them, except for one (see below), they are immediately cast to a uint. This happens explicitly e.g. [here](https://github.com/bitcoin/bitcoin/blob/9fcdb9f3a044330d3d7515fa35709102c98534d2/src/txmempool.cpp#L123), or implicitly by passing it to a fn/ctor/variable that expects a uint or size_t (e.g. [here](https://github.com/bitcoin/bitcoin/blob/9fcdb9f3a044330d3d7515fa35709102c98534d2/src/node/interfaces.cpp#L662-L664) or [here](https://github.com/bitcoin/bitcoin/blob/9fcdb9f3a044330d3d7515fa35709102c98534d2/src/node/interfaces.cpp#L650-L651) or [here](https://github.com/bitcoin/bitcoin/blob/9fcdb9f3a044330d3d7515fa35709102c98534d2/src/validation.cpp#L427-L431)).

The only place where we actually use the signedness of any of the members is here: https://github.com/bitcoin/bitcoin/blob/9fcdb9f3a044330d3d7515fa35709102c98534d2/src/init.cpp#L1418 
I don't think this implicit conversion is problematic.

For that reason, I would argue that updating the `MemPoolLimits` members to be `uint64_t` instead of `int64_t` (which is necessary to avoid compiler warnings that we're comparing uint with int in all of the places outlined the diff [here](https://github.com/bitcoin/bitcoin/pull/26103#discussion_r975593764)) is - except for the one case - not a regression, and is the most straightforward implementation for the goal of this PR. I currently have no objection to what is being proposed in https://github.com/bitcoin/bitcoin/pull/23962, but I feel like changing these members and all subsequent call sites to signed integers is orthogonal and would unnecessarily introduce complexity and controversy for this PR. As such, my preference would be to not do it here.

What do you think?

glozow replied:
Ah I see, thank you very much for going through the usage! I agree it seems appropriate to defer to a later PR.

hebasto replied:
3611c2bd710d147fbed815180e512f2af0061e83, approach NACK.

I do not agree that switching to unsigned type is an improvement. Especially for variables involved in arithmetic and comparison operations.

For example, see https://www.aristeia.com/Papers/C++ReportColumns/sep95.pdf

stickies-v replied:
Thanks for your feedback hebasto. Even though the `MemPoolLimits` members are almost always cast to an unsigned int so I think it'd be clearer to have everything be in the type in which it's actually used, I understand your concern that you don't want to regress the interface to avoid future PRs building on an unsigned `MemPoolLimits` interface.

As such, I think the most straightforward way to move this PR forward and preserve as much review time as possible is to remove the commit that made the `MemPoolLimits` members unsigned and introduce `static_cast<uint64_t>` wherever necessary. This way the interface remains signed and behaviour remains unchanged (including arithmetic with unsigned integers) to how it was before this PR.

glozow reviewed (commented):
Concept ACK to using the `Limits` struct instead of a list of ints. I think this is more readable, especially the `static NoLimits` when we just want ancestors calculated.

aureleoules commented on src/txmempool.h:
```diff
@@ -670,36 +667,31 @@ class CTxMemPool
-                            uint64_t limitAncestorSize,
-                            uint64_t limitDescendantCount,
-                            uint64_t limitDescendantSize,
+                            Limits limits,
```
```suggestion
                            const Limits& limits,
```

aureleoules commented on src/txmempool.cpp:
```diff
@@ -187,10 +187,7 @@ bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
-                                                  uint64_t limitAncestorSize,
-                                                  uint64_t limitDescendantCount,
-                                                  uint64_t limitDescendantSize,
+                                                  CTxMemPool::Limits limits,
```
```suggestion
                                                  const CTxMemPool::Limits& limits,
```

aureleoules commented on src/txmempool.cpp:
```diff
@@ -233,10 +230,7 @@ bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
-                                    uint64_t limitAncestorSize,
-                                    uint64_t limitDescendantCount,
-                                    uint64_t limitDescendantSize,
+                                    CTxMemPool::Limits limits,
```
```suggestion
                                    const CTxMemPool::Limits& limits,
```

aureleoules reviewed (commented):
Concept ACK

fanquake commented:
Concept ACK

stickies-v reviewed (commented):
Force pushed to address review feedback and remove a leftover whitespace. Thank you for the quick feedback everyone!

Main changes:
- Introduce `CPFPCarveOutLimits()` helper function in `mempool_limits.h`. Currently not used by anyone else, but a bit tidier.
- pass `Limits` by reference instead of value

aureleoules commented:
you commited `compile_commands.json`

stickies-v commented:
> you commited compile_commands.json

whoops sorry, added that to global gitignore

Riahiamirreza commented:
As far as I could understand the PR Concept ACK. I think this PR is relatively easy to understand and putting the `good-first-review` label on it can make sense IMO.

stickies-v commented:
Force pushed to fix the failing `./test/functional/mempool_package_onemore.py` test. In the last refactoring ([c5e5952](https://github.com/bitcoin/bitcoin/commit/c5e595250e26491b4dfbccab83cd7c6fa77df331) -> [c17d468](https://github.com/bitcoin/bitcoin/commit/c17d4683f8cb30be8a6f1908afdc346ba3e3b30e)) I should have passed `m_limits` when calling `CPFPCarveOutLimits()` because it was modified earlier in `PreChecks()` already. Also removed the argument-less overload of `CPFPCarveOutLimits()` since it's now not used anywhere anymore.

glozow commented on src/validation.cpp:
```diff
@@ -424,11 +424,18 @@ namespace {
+        m_view(&m_dummy),
+        m_viewmempool(&active_chainstate.CoinsTip(), m_pool),
+        m_active_chainstate(active_chainstate),
+        m_limits{
+            .ancestor_count = m_pool.m_limits.ancestor_count,
+            .ancestor_size_vbytes = m_pool.m_limits.ancestor_size_vbytes,
+            .descendant_count = m_pool.m_limits.descendant_count,
+            .descendant_size_vbytes = m_pool.m_limits.descendant_size_vbytes
+        }
```
nit in 55b8e6f6cb:

Why not just use default copy?
```suggestion
        m_limits{m_pool.m_limits}
```

stickies-v replied:
That's definitely how a sane person would do it yes, woops. Thank you - updated!

glozow reviewed (commented):
code review ACK, just 1-2 nits

stickies-v commented on src/kernel/mempool_limits.h:
```diff
@@ -17,13 +17,33 @@ namespace kernel {
+    }
+
+    /** Modify the current mempool limits to reflect what is allowed under CPFP carve out policy. */
+    static MemPoolLimits CPFPCarveOutLimits(const MemPoolLimits& current_limits)
```
Okay that makes sense, I didn't have a strong view either way - no problem at all. I've reverted it back to the previous version (calling it `cpfp_carve_out_limits` this time) and then it can be refactored out in a follow-up.

stickies-v reviewed (commented):
Force pushed to incorporate glozow's feedback - thank you for the review!
- using copy constructor instead of designated initializer
- removed the carved-out `CPFPCarveOutLimits()` function (~revert to previous version)

glozow reviewed (commented):
ACK ae3a5f33f9ba0037bb01d4cb2d6713c8e2abd2ed

aureleoules reviewed (approved):
ACK ae3a5f33f9ba0037bb01d4cb2d6713c8e2abd2ed

hebasto commented:
Concept ACK.

stickies-v commented:
Force pushed to [address](https://github.com/bitcoin/bitcoin/pull/26103#discussion_r986909661) hebasto's [concerns](https://github.com/bitcoin/bitcoin/pull/26103#discussion_r986641966) regarding making the `MemPoolLimits` interface unsigned, which goes counter to the direction we want to take.

Note: this introduces potential for follow-up improvements in `CTxMemPool::CalculateAncestorsAndCheckLimits()` where some of the `static_cast`s can be removed by updating the function signature to use `int64_t` instead of `size_t` when `CTxMemPoolEntry::GetCountWithDescendants()` and `CTxMemPoolEntry::GetSizeWithDescendants()` are updated to return an `int64_t`. I'd prefer to minimize the

stickies-v commented:
Force pushed to rebase to fix failing Win64 native CI (unrelated to PR).

aureleoules reviewed (approved):
re-ACK d0bfe6869855dcf112a7da640e2e8ad648f82bbd.
Since my last review: 
- rolled back from `uint64_t` to `int64_t` for `descendant_count`, `descendant_size_vbytes` and `no_limit`.
- added static casts to compare these values against `uint64_t` types.

I also verified the values being checked against are of type `size_t` or `uint64_t`.

hebasto commented:
Approach ACK d0bfe6869855dcf112a7da640e2e8ad648f82bbd.

hebasto commented on src/kernel/mempool_limits.h:
```diff
@@ -24,6 +24,15 @@ struct MemPoolLimits {
+    /**
+     * @return MemPoolLimits with all the limits set to the maximum
+     */
+    static MemPoolLimits NoLimits()
```
67a4932a2ca1131ea04a011d67b3f3c2da727b5f
```suggestion
    static constexpr MemPoolLimits NoLimits()
```

stickies-v replied:
No strong view. It's currently not called at compile time so perhaps a bit premature, but simultaneously it seems unlikely we'll ever want to not guarantee this fn to be `constexpr` so might as well add it already. Updated.

stickies-v replied:
As @hebasto kindly explained offline (comparing https://godbolt.org/z/78r3sY4TW with https://godbolt.org/z/TfvKKvdM9), using `constexpr` here does indeed have benefits as the compiler is able to replace the `NoLimits()` function calls with constants. Thank you!

hebasto commented on src/test/mempool_tests.cpp:
```diff
@@ -203,7 +203,7 @@ BOOST_AUTO_TEST_CASE(MempoolIndexingTest)
     CTxMemPool::setEntries setAncestorsCalculated;
     std::string dummy;
-    BOOST_CHECK_EQUAL(pool.CalculateMemPoolAncestors(entry.Fee(2000000LL).FromTx(tx7), setAncestorsCalculated, 100, 1000000, 1000, 1000000, dummy), true);
+    BOOST_CHECK_EQUAL(pool.CalculateMemPoolAncestors(entry.Fee(2'000'000LL).FromTx(tx7), setAncestorsCalculated, CTxMemPool::Limits::NoLimits(), dummy), true);
```
Why has test been modified?

If it is required, it would be better to separate such changes into their own commit to ensure that tests pass with no changes in testable code, no?

glozow replied:
See https://github.com/bitcoin/bitcoin/pull/26103#discussion_r972865346

separate commit would be fine

stickies-v replied:
I've split it out into separate commit with glozow's rationale in the commit message.

stickies-v committed:
test: use NoLimits() in MempoolIndexingTest

The (100, 1000000, 1000, 1000000) limits are arbitrarily high and
don't restrict anything, they are just meant to calculate ancestors
properly. Using NoLimits() makes this intent more clear and simplifies
the code.

stickies-v committed:
docs: improve docs where MemPoolLimits is used

stickies-v commented:
Force pushed to address hebasto's review feedback - thank you!
- Split out changes to `test/mempool_tests.cpp` into separate commit
- Added commit with improvements to Doxygen comments
- Made `MemPoolLimits::NoLimits()` `constexpr`

hebasto commented on src/txmempool.cpp:
```diff
@@ -203,14 +200,14 @@ bool CTxMemPool::CalculateAncestorsAndCheckLimits(size_t entry_size,
             return false;
-        } else if (totalSizeWithAncestors > limitAncestorSize) {
-            errString = strprintf("exceeds ancestor size limit [limit: %u]", limitAncestorSize);
+        } else if (totalSizeWithAncestors > static_cast<uint64_t>(limits.ancestor_size_vbytes)) {
```
nit: 
```suggestion
        } else if (totalSizeWithAncestors > static_cast<size_t>(limits.ancestor_size_vbytes)) {
```

I beg a pardon for being pedantic :) Feel free to ignore this nit.

stickies-v replied:
Since this is a refactor PR, I'll leave this as is. `limitAncestorSize` was a `uint64_t` before, so better to change this in a future PR imo. I'll leave this comment open for visibility.

hebasto reviewed (approved):
ACK 33b12e5df6aa14344dd084e0c6f2d34158fca383, I have reviewed the code and it looks OK, I agree it can be merged.

glozow reviewed (commented):
reACK 33b12e5df6aa14344dd084e0c6f2d34158fca383
